src/marked.js

```diff
@@ -139,45 +139,99 @@ marked.defaults = defaults;
  */
 
 marked.use = function(extension) {
-  const opts = merge({}, extension);
-  if (extension.renderer) {
-    const renderer = marked.defaults.renderer || new Renderer();
-    for (const prop in extension.renderer) {
-      const prevRenderer = renderer[prop];
-      renderer[prop] = (...args) => {
-        let ret = extension.renderer[prop].apply(renderer, args);
-        if (ret === false) {
-          ret = prevRenderer.apply(renderer, args);
+  if (!Array.isArray(extension)) { // Wrap in array if not already to unify processing
+    extension = [extension];
+  }
+
+  const opts = merge({}, ...extension);
+  const extensions = marked.defaults.extensions || { renderers: {}, walkableTokens: {} };
+  let hasExtensions;
+
+  extension.forEach((pack) => {
+    // ==-- Parse "addon" extensions --== //
+    if (pack.extensions) {
+      hasExtensions = true;
+      pack.extensions.forEach((ext) => {
+        if (ext.renderer && ext.name) { // Renderers must have 'name' property
+          extensions.renderers[ext.name] = ext.renderer;
         }
-        return ret;
-      };
+        if (ext.walkableTokens && ext.name) { // walkableTokens must have 'name'
+          extensions.walkableTokens[ext.name] = ext.walkableTokens;
+        }
+        if (ext.tokenizer && ext.level) { // Tokenizers must have 'level' property
+          if (extensions[ext.level]) {
+            extensions[ext.level].push(ext.tokenizer);
+          } else {
+            extensions[ext.level] = [ext.tokenizer];
+          }
+          if (ext.start) { // Function to check for start of token
+            if (ext.level === 'block') {
+              if (extensions.startBlock) {
+                extensions.startBlock.push(ext.start);
+              } else {
+                extensions.startBlock = [ext.start];
+              }
+            } else if (ext.level === 'inline') {
+              if (extensions.startInline) {
+                extensions.startInline.push(ext.start);
+              } else {
+                extensions.startInline = [ext.start];
+              }
+            }
+          }
+        }
+      });
     }
-    opts.renderer = renderer;
-  }
-  if (extension.tokenizer) {
-    const tokenizer = marked.defaults.tokenizer || new Tokenizer();
-    for (const prop in extension.tokenizer) {
-      const prevTokenizer = tokenizer[prop];
-      tokenizer[prop] = (...args) => {
-        let ret = extension.tokenizer[prop].apply(tokenizer, args);
-        if (ret === false) {
-          ret = prevTokenizer.apply(tokenizer, args);
+
+    // ==-- Parse "overwrite" extensions --== //
+    if (pack.renderer) {
+      const renderer = marked.defaults.renderer || new Renderer();
+      for (const prop in pack.renderer) {
+        const prevRenderer = renderer[prop];
+        // Replace renderer with func to run extension, but fall back if fail
```

Comment:
It falls back if the renderer returns `false` not if it throws an error.

```suggestion
        // Replace renderer with func to run extension, but fall back if false
```